sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java

```diff
@@ -129,13 +129,14 @@ private void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certifi
         try {
             trustManagerConsumer.checkTrusted(this);
         } catch (CertificateException e) {
-            boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
-            if (shouldBeTrusted) {
-                addCertificates(Collections.singletonList(chain[0]));
-                return;
+            synchronized (this) {
```

Comment:
Ah, yes I see it what you are trying to do and it is getting better, however multiple threads are able to call checkTrusted while they need to wait before entering the exception block. What will happen when user trust the untrested certificate but another thread is having the same certificate. The user already accepted the certificate, however the other thread was waiting till the catch block was accesible again and will prompt the user with the same question for the same certificate.

What do you think of addind the syncronize block to the method. It will therefor allow only 1 method at a time to enter and do the certificate checking and if is is untrusted the user will be prompt and if trusted, the another thread with the same certificate will pass instantly as it will call the checktrusted after the trustmanager is getting updated. Maybe something like this:

```java
private synchronized void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certificate[] chain, String authType) throws CertificateException {
    try {
        trustManagerConsumer.checkTrusted(this);
    } catch (CertificateException e) {
        boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
        if (shouldBeTrusted) {
            addCertificates(Collections.singletonList(chain[0]));
        } else {
            throw e;
        }
    }
}
```

What do you think, does it make sense?

Reply:
> What do you think of addind the syncronize block to the method
> What do you think, does it make sense?

If think this is fine, and pretty safe, since this is more or less what is done on IntelliJ [side](https://github.com/JetBrains/intellij-community/blob/f3ac42aab02f84a92afbfc1fb48ae15906e07297/platform/platform-api/src/com/intellij/util/net/ssl/ConfirmingTrustManager.java#L216). It might be a performance issue in situation where there are many concurrent requests, that's why I suggested the [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking). But for us, both will be fine, since we don't expect a high frequency of requests.

Reply:
I didn't know about the double-checked locking. I am able to understand what you did and I think it is a nice solution. My own code snippet causes more overhead and is less performing and can have performance issues for some end-user.

I would like to merge your PR into this branch, however it has some merge conflicts. Can you resolve those on your side and push your commit?

Reply:
Done